File: api/routes/movements.go
package routes

import (
	"transaction-tracker/api/handler"
	"transaction-tracker/api/models"
)

func MovementsRoutes(h *handler.MovementHandler) []models.Route {
	return []models.Route{
		{
			Endpoint:    "/movements",
			Method:      models.GET,
			HandlerFunc: h.GetMovements,
			ApiVersion:  API_VERSION,
		},
		{
			Endpoint:    "/movements/:id",
			Method:      models.GET,
			HandlerFunc: h.GetMovementByID,
			ApiVersion:  API_VERSION,
		},
		{
			Endpoint:    "/movements",
			Method:      models.POST,
			HandlerFunc: h.CreateMovement,
			ApiVersion:  API_VERSION,
		},
		{
			Endpoint:    "/movements/:id",
			Method:      models.DELETE,
			HandlerFunc: h.DeleteMovement,
			ApiVersion:  API_VERSION,
		},
		{
			Endpoint:    "/movements/years/:year",
			Method:      models.GET,
			HandlerFunc: h.GetMovementsByYear,
			ApiVersion:  API_VERSION,
		},
		{
			Endpoint:    "/movements/years/:year/months/:month",
			Method:      models.GET,
			HandlerFunc: h.GetMovementsByMonth,
			ApiVersion:  API_VERSION,
		},
	}
}

File: frontends/transactions-dashboard/src/infrastructure/repositories/movements.ts
import type {
  IMovementsRepository,
  Movement,
  MovementByYear,
  MovementRequest,
  MovementsResponse
} from "@/core/entities/Movement";
import { createFetchClient, API_BASE_URL } from "../http/fetchClient";
import { ExtendedBody, FetchClient } from "@/core/entities/FetchClient";
import { MovementByMonth } from "@/core/entities/Movement.d";

export class MovementsRepository implements IMovementsRepository {
  private client: FetchClient;

  constructor() {
    this.client = createFetchClient(API_BASE_URL);
  }

  createMovement(movement: MovementRequest): Promise<Movement> {
    return this.client<Movement>(`/movements`, {
      method: "POST",
      body: movement
    });
  }

  getMovements(page: number, intitutionIDs: string[]): Promise<MovementsResponse> {
    return this.client<MovementsResponse>(`/movements?page=${page}&limit=20&institution_ids=${intitutionIDs.join(",")}`);
  }

  getMovementsByYear(year: number, intitutionIDs: string[]): Promise<MovementByYear> {
    return this.client<MovementByYear>(`/movements/years/${year}?institution_ids=${intitutionIDs.join(",")}`);
  }

  getMovementsByMonth(year: number, month: number, intitutionIDs: string[]): Promise<MovementByMonth> {
    month += 1;

    return this.client<MovementByMonth>(
      `/movements/years/${year}/months/${month}?institution_ids=${intitutionIDs.join(",")}`
    );
  }
}
